static/usr/share/man/man8/hcron-scheduler.8
.TH HCRON-SCHEDULER "8" "2020" "hcron 1.x" ""
.SH NAME
hcron-scheduler \- Server which schedules and executes hcron events
.SH SYNOPSIS
.B hcron-scheduler

.SH DESCRIPTION
The scheduler loads event definitions set up for the local host and
executes commands accordingly. When run by the root user, event
definitions of those users listed in the hcron.allow file are managed;
otherwise, only the events of the current user are managed. The standard
operation is to run as root.

Certain operational information is logged to a specific log file or to
the syslog service.

.SH METHOD OF OPERATION
.PP
Each minute, hcron-scheduler performs a number of checks in order to
keep up to date:

.IP \[bu] 2
A change in the modification time of the hcron.conf file forces a full
restart of the server.

.IP \[bu] 2
A change in the modification time of the hcron.allow file forces a
reload of the list of allowed users, the event information of
non-existent users is discarded, and the event definitions of new users
is loaded.

.IP \[bu] 2
The hcron spool directory is checked for files. For each file that is
found, the event information of the user name owning the file is
reloaded as appropriate.

.SH HCRON.ALLOW FILE
.PP
The hcron.allow file contains a list of all users for which hcron-scheduler
will load event definitions and schedule. Empty lines and those starting
with # are ignored.

.SH HCRON.CONF FILE
.PP
The hcron.conf file contains configuration settings for hcron-scheduler.
It takes the form of a simplified (simple object types) Python dictionary
with the following keys:

.TP
.B allow_localhost
Boolean indicating whether or not event commands may be done on the
localhost, as specified by the host setting of the event definition.
This restriction applies to any address which resolves to the local
host.

.TP
.B allow_root_events
Boolean indicating whether events belonging to root will be run.
Default is False.

.TP
.B command_spawn_timeout
Maximum time allowed for a command to be spawned. This does not limit
the time allowed for a command to execute, but serves to limit spawned
processes on the hcron machine.

.TP
.B events_base_path
Path below which to search for user event definitions, following the
pattern <events_base_path>/<username>. If undefined or None, user
definitions are loaded from the user's home (~<username>).

.TP
.B log_path
Path of the log file, when use_syslog is False. A relative path is
prepended with /var/log.

.TP
.B max_activated_events
Maximum number of events that can be activated (spawning) at one time.
Chained events count as 1.

.TP
.B max_chain_events
Maximum number of chain events from a single event. The first event counts
as 1.

.TP
.B max_events_per_user
Maximum number of events allowed per user.

.TP
.B max_hcron_tree_snapshot_size
Maximum size of the hcron event tree snapshot created with "hcron reload".

.TP
.B names_to_ignore_regexp
Regular expression matching event (and directory) names to ignore when
loading events. The default is to ignore hidden files (starting with .)
and names ending with ~ (commonly used to name backup or temporary
files when editing).

.TP
.B server_name
Name of the server for which events are being scheduled. Default is the
fqdn, but may be overridden with this setting.

.TP
.B smtp_server
Host name of an SMTP server willing to accept connections. If not
specified, localhost is used.

.TP
.B test_net_delay
Time to wait between retries of the test of the naming service.

.TP
.B test_net_retry
Number of times to retry to determine if the naming service is
available.

.TP
.B test_net_username
A username of a network account (i.e., for which user information is
served by YP/NIS, LDAP, etc.). This is used to test the status of said
service in case a lookup for user information fails which could be
because the service is inaccessible or the user does not exist.

.TP
.B use_syslog
Boolean indicating whether or not to send the logging information to
the syslog daemon.

.PP
As in Python, empty lines and those starting with # are ignored.

.SH HCRON SPOOL DIRECTORY
.PP
The hcron spool directory is accessible to all users within which signal
files can be created using hcron-reload. At the start of each scheduling
period, the signal files are checked. For each signal file, the event
definition files (the snapshot) of its owner are reloaded. Then, the
signal files are removed. The worst case reload wait is just short of
two scheduling periods.

.SH EXAMPLES
.PP
A default hcron.conf configuration:

.RS
.nf
# hcron.conf
{
    "allow_localhost": False,
    "command_spawn_timeout": 15,
    "events_base_path": None,
    "log_path": "hcron.log",
    "max_events_per_user": 25,
    "names_to_ignore_regexp": "(\..*)|(.*~$)",
    "smtp_server": "localhost",
    "use_syslog": False,
}
.fi
.RE

.SH SIGNALS
hcron-scheduler reacts to the following signals by:
.RS
.nf
kill -<signal> `cat /var/run/hcron.pid`
.fi
.RE

.IP SIGHUP
Reload all configuration files and event definitions.

.IP SIGTERM
Terminate.

.SH FILES
.IP /etc/hcron/hcron.allow
See HCRON.ALLOW above.

.IP /etc/hcron/hcron.conf
See HCRON.CONF above.

.IP /var/lib/hcron/event_lists/<username>
See hcron (hcron reload).
 
.IP /var/lib/hcron/trees/<username>
See hcron (hcron reload).
 
.IP /var/log/hcron/hcron.log
Default location of the log file (when use_syslog is False).

.IP /var/run/hcron.pid
Pid file for the currently running hcron-scheduler. This is managed by
hcron-scheduler itself.

.SH SEE ALSO
hcron(1)

.SH AUTHOR
Written by John Marshall.

.SH COPYRIGHT
Copyright \(co 2008-2020 Environment Canada.
.br
This is free software.  You may redistribute copies of it under the terms of
the GNU General Public License <http://www.gnu.org/licenses/gpl.html>.
There is NO WARRANTY, to the extent permitted by law.


static/usr/lib/hcron/hcron-scheduler.py
#! /usr/bin/env python3
#! /usr/bin/env python2
#
# hcron-scheduler.py

# GPL--start
# This file is part of hcron
# Copyright (C) 2008-2019 Environment/Environnement Canada
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; version 2
# of the License.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
# GPL--end

"""This is the backend event scheduler. It is typically run by root
although it can be run by individual users, but with reduced
functionality (i.e., it only handlers the user's own events).
"""

# secure by restricting sys.path to /usr and first path entry
import sys
firstPath = sys.path[0]
sys.path = [ path for path in sys.path if path.startswith("/usr") ]
sys.path.insert(0, firstPath)
del firstPath

# system imports
import os.path
import signal
from sys import stderr
import tempfile
import traceback

# app imports
from hcron import globs
from hcron import library
from hcron.constants import *
from hcron.event import EventListList
from hcron.file import PidFile
from hcron.logger import *
from hcron.server import Server, setup
from hcron.trackablefile import AllowFile, SignalDir

def dump_signal_handler(num, frame):
    log_message("info", "received signal to dump.")
    signal.signal(num, dump_signal_handler)

    try:
        library.makedirs(HCRON_DUMPDIR_BASE, 0o700)
        dumpdir = tempfile.mkdtemp(dir=HCRON_DUMPDIR_BASE)
        elsdumpdir = os.path.join(dumpdir, "event_lists")
        library.makedirs(elsdumpdir, 0o700)
    except:
        # something went wrong!
        return

    # config
    try:
        globs.configfile.dump(os.path.join(dumpdir, "hcron.conf"))
    except Exception:
        pass

    # allowed users
    try:
        globs.allowfile.dump(os.path.join(dumpdir, "hcron.allow"))
    except Exception:
        pass

    # event list
    try:
        ell = globs.eventlistlist
        for username in globs.allowfile.get():
            el = ell.eventlists.get(username)
            if el:
                el.dump(elsdumpdir)
    except:
        pass

    # threadpool information
    try:
        tp = globs.server.jobq.tp
        l = []
        l.append("ndone (%s)" % tp.get_ndone())
        l.append("nrunning (%s)" % tp.get_nrunning())
        l.append("nwaiting (%s)" % tp.get_nwaiting())
        l.append("nworkers (%s)" % tp.get_nworkers())
        l.append("\nrunning:")
        l.append("\n".join(["%s" % x for x in tp.runs]))
        open(os.path.join(dumpdir, "threadpool"), "w+").write("\n".join(l))
    except:
        pass

def reload_signal_handler(num, frame):
    log_message("info", "received signal to reload.")
    signal.signal(num, reload_signal_handler)
    globs.eventlistlist.load(globs.allowfile.get())

def quit_signal_handler(num, frame):
    log_message("info", "received signal to exit.")
    globs.pidfile.remove()
    sys.exit(0)

def print_usage():
    d = {
        "progname": os.path.basename(sys.argv[0])
    }
    print("""\
usage: %(progname)s [--immediate]
       %(progname)s -h|--help

This program loads a collection of event files from one or more users
and executes commands according their defined schedulues. When run as
root, event files are read from registered users (listed in hcron.allow)
for the local host. Otherwise, this is done for the current user, only.

Options:
--immediate         Forces the scheduling of events to be done
                    immediately (i.e., now, the current interval)
                    rather than wait for the next interval.""" % d)

if __name__ == "__main__":
    try:
        immediate = False

        args = sys.argv[1:]
        while args:
            arg = args.pop(0)
            if arg == "--immediate":
                immediate = True
            elif arg in ["-h", "--help"]:
                print_usage()
                sys.exit(0)
            else:
                raise Exception()
    except SystemExit:
        raise
    except:
        stderr.write("error: bad/missing argument\n")
        sys.exit(1)

    try:
        setup()

        globs.remote_execute_enabled = True
        globs.email_notify_enabled = True

        setup_logger()

        globs.allowfile = AllowFile(HCRON_ALLOW_PATH)
        globs.signaldir = SignalDir(HCRON_SIGNAL_DIR)
        globs.eventlistlist = EventListList(globs.allowfile.get())

        signal.signal(signal.SIGHUP, reload_signal_handler)
        signal.signal(signal.SIGUSR1, dump_signal_handler)
        signal.signal(signal.SIGTERM, quit_signal_handler)
        signal.signal(signal.SIGQUIT, quit_signal_handler)
        ###signal.signal(signal.SIGCHLD, signal.SIG_IGN)   # we don't care about children/zombies

        library.serverize()  # don't catch SystemExit

        globs.server = Server()
        globs.pidfile = PidFile(HCRON_PID_FILE_PATH)
        globs.pidfile.create()

        log_start()
        globs.server.run(immediate=immediate)
    except Exception as detail:
        log_message("error", "unexpected error (%s)." % detail)

    try:
        globs.pidfile.remove()
        log_exit()
    except:
        pass

    sys.exit(1)
